### PR TITLE
LCS-29: telemetry & logging

### PR DESCRIPTION
**Pull request by quasisamurai**
Here are `debug` logging + [telemetry](https://docs.cosmos.network/master/core/telemetry.html).

if you think that we need to add more tele[metrics] or concrete logging cases - please be free to write your suggestion

also, there are a couple of issues:
-  do we need on initialization logging?
- what retention for prom-metrics we prefer? infinite?

### CODE REVIEW

File: x/interchainqueries/keeper/abci.go [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"strconv"
+	"time"
 
 	"github.com/neutron-org/neutron/x/interchainqueries/types"
 )
 
 // EndBlocker of interchainquery module
 func (k Keeper) EndBlocker(ctx sdk.Context) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
```
Comment (zavgorodnii) [inline x/interchainqueries/keeper/abci.go:14]: please add a \n after the defer statement throughout the PR

File: x/interchainqueries/keeper/abci.go [hunk 2]
```diff
@@ -26,18 +28,18 @@ func (k Keeper) EndBlocker(ctx sdk.Context) {
...
 			)
-
 			events = append(events, event)
 			registeredQuery.LastEmittedHeight = uint64(ctx.BlockHeight())
 			if err := k.SaveQuery(ctx, registeredQuery); err != nil {
-				k.Logger(ctx).Error("failed to save query: %w", err)
+				k.Logger(ctx).Error("Failed to save query", "error", err)
 			}
-
+			k.Logger(ctx).Debug("Interchainsquery event successfully added to events list", "id", registeredQuery)
```
Comment (zavgorodnii) [inline x/interchainqueries/keeper/abci.go]: Interchainsquery?

File: x/interchaintxs/keeper/msg_server.go [hunk 3]
```diff
@@ -32,23 +33,28 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 }
 
 func (k Keeper) RegisterInterchainAccount(goCtx context.Context, msg *ictxtypes.MsgRegisterInterchainAccount) (*ictxtypes.MsgRegisterInterchainAccountResponse, error) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), LabelRegisterInterchainAccount)
 	ctx := sdk.UnwrapSDKContext(goCtx)
+	k.Logger(ctx).Debug("Registering interchain account", "connection_id", msg.ConnectionId, "from_address", msg.FromAddress, "interchain_accountt_id", msg.InterchainAccountId)
 
 	icaOwner, err := types.NewICAOwner(msg.FromAddress, msg.InterchainAccountId)
 	if err != nil {
+		k.Logger(ctx).Error("Registering interchain account failed: failed to create LabelRegisterInterchainAccount", "error", err)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go]: Please stick to the same pattern throughout the PR:

```
k.Logger(ctx).Error("RegisterInterchainAccount: failed to XXX", "error", err)
```

Also — what does `failed to create LabelRegisterInterchainAccount` mean here? Shouldn't it be something like

```
k.Logger(ctx).Debug("RegisterInterchainAccount: failed to" create NewICAOwner, "connection_id", msg.ConnectionId, "from_address", msg.FromAddress, "interchain_accountt_id", msg.InterchainAccountId)
```

File: x/interchaintxs/keeper/msg_server.go [hunk 4]
```diff
@@ -32,23 +33,28 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
...
 		return nil, sdkerrors.Wrap(err, "failed to create ICA owner")
 	}
 
 	if err := k.icaControllerKeeper.RegisterInterchainAccount(ctx, msg.ConnectionId, icaOwner.String()); err != nil {
-		k.Logger(ctx).Error("failed to create RegisterInterchainAccount:", "error", err, "owner", icaOwner.String(), "connection_id", msg.ConnectionId)
-		return nil, sdkerrors.Wrap(err, "failed to RegisterInterchainAccount")
+		k.Logger(ctx).Error("Register interchain account failed: failed to create LabelRegisterInterchainAccount:", "error", err, "owner", icaOwner.String(), "msg", &msg)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go]: Same here, what is `LabelRegisterInterchainAccount`? It should be:

```
k.Logger(ctx).Error("RegisterInterchainAccount: failed to create RegisterInterchainAccount:", "error", err, "owner", icaOwner.String(), "msg", &msg)
```

File: x/interchaintxs/keeper/msg_server.go (hunk 3 above)
Comment (quasisamurai) [reply x/interchaintxs/keeper/msg_server.go]: >`LabelRegisterInterchainAccount ` 
refactor rename did this

File: x/interchaintxs/keeper/msg_server.go [hunk 5]
```diff
@@ -32,23 +33,28 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
...
+		return nil, sdkerrors.Wrap(err, "failed to LabelRegisterInterchainAccount")
 	}
 
 	return &ictxtypes.MsgRegisterInterchainAccountResponse{}, nil
 }
 
 func (k Keeper) SubmitTx(goCtx context.Context, msg *ictxtypes.MsgSubmitTx) (*ictxtypes.MsgSubmitTxResponse, error) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), LabelSubmitTx)
 	ctx := sdk.UnwrapSDKContext(goCtx)
+	k.Logger(ctx).Debug("Submitting tx", "connection_id", msg.ConnectionId, "from_address", msg.FromAddress, "interchain_accountt_id", msg.InterchainAccountId)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go]: Please just use the method name:

```
k.Logger(ctx).Debug("SubmitTx", "connection_id", msg.ConnectionId, "from_address", msg.FromAddress, "interchain_accountt_id", msg.InterchainAccountId)
```

File: x/interchaintxs/keeper/msg_server.go [hunk 6]
```diff
@@ -57,31 +63,31 @@ func (k Keeper) SubmitTx(goCtx context.Context, msg *ictxtypes.MsgSubmitTx) (*ic
 
 	portID, err := icatypes.NewControllerPortID(icaOwner.String())
 	if err != nil {
-		k.Logger(ctx).Error("failed to create NewControllerPortID:", "error", err, "owner", icaOwner)
+		k.Logger(ctx).Error("SubmitTx failed: failed to create NewControllerPortID:", "error", err, "owner", icaOwner)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go]: Please stick to:

```
k.Logger(ctx).Error("SubmitTx: failed to create NewControllerPortID:", "error", err, "owner", icaOwner)
```

To consecutive "failed" statements are redundant here.

File: x/interchaintxs/keeper/msg_server.go [hunk 7]
```diff
@@ -32,23 +33,28 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
...
 
 	return &ictxtypes.MsgRegisterInterchainAccountResponse{}, nil
 }
 
 func (k Keeper) SubmitTx(goCtx context.Context, msg *ictxtypes.MsgSubmitTx) (*ictxtypes.MsgSubmitTxResponse, error) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), LabelSubmitTx)
 	ctx := sdk.UnwrapSDKContext(goCtx)
+	k.Logger(ctx).Debug("Submitting tx", "connection_id", msg.ConnectionId, "from_address", msg.FromAddress, "interchain_accountt_id", msg.InterchainAccountId)
 
 	icaOwner, err := types.NewICAOwner(msg.FromAddress, msg.InterchainAccountId)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go:61]: should  be logged?

File: x/interchaintxs/keeper/msg_server.go (hunk 3 above)
Comment (zavgorodnii) [reply x/interchaintxs/keeper/msg_server.go]: Also this should not be at ERROR level. The error here can only be triggered by malformed user input, this is not an internal Neutron error that we should inform the validator about

File: x/interchaintxs/keeper/msg_server.go (hunk 4 above)
Comment (zavgorodnii) [reply x/interchaintxs/keeper/msg_server.go]: This should also not be at ERROR level. You pass unchecked connection ID from user input, which can look like `&*9r4`; this is not necessary an internal Neutron error.

File: x/interchaintxs/keeper/msg_server.go (hunk 6 above)
Comment (zavgorodnii) [reply x/interchaintxs/keeper/msg_server.go]: DEBUG

File: x/interchaintxs/keeper/msg_server.go [hunk 8]
```diff
@@ -57,31 +63,31 @@ func (k Keeper) SubmitTx(goCtx context.Context, msg *ictxtypes.MsgSubmitTx) (*ic
...
 	if err != nil {
-		k.Logger(ctx).Error("failed to create NewControllerPortID:", "error", err, "owner", icaOwner)
+		k.Logger(ctx).Error("SubmitTx failed: failed to create NewControllerPortID:", "error", err, "owner", icaOwner)
 		return nil, sdkerrors.Wrap(err, "failed to create NewControllerPortID")
 	}
 
 	channelID, found := k.icaControllerKeeper.GetActiveChannelID(ctx, msg.ConnectionId, portID)
 	if !found {
-		k.Logger(ctx).Error("failed to GetActiveChannelID", "connection_id", msg.ConnectionId, "port_id", portID)
+		k.Logger(ctx).Error("SubmitTx failed: failed to GetActiveChannelID", "connection_id", msg.ConnectionId, "port_id", portID)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go]: DEBUG

File: x/interchaintxs/keeper/msg_server.go [hunk 9]
```diff
@@ -57,31 +63,31 @@ func (k Keeper) SubmitTx(goCtx context.Context, msg *ictxtypes.MsgSubmitTx) (*ic
...
 		return nil, sdkerrors.Wrapf(icatypes.ErrActiveChannelNotFound, "failed to GetActiveChannelID for port %s", portID)
 	}
 
 	chanCap, found := k.scopedKeeper.GetCapability(ctx, host.ChannelCapabilityPath(portID, channelID))
 	if !found {
-		k.Logger(ctx).Error("failed to GetCapability", "connection_id", msg.ConnectionId, "port_id", portID, "channel_id", channelID)
+		k.Logger(ctx).Error("SubmitTx failed: failed to GetCapability", "connection_id", msg.ConnectionId, "port_id", portID, "channel_id", channelID)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go]: DEBUG

File: x/interchaintxs/keeper/msg_server.go [hunk 10]
```diff
@@ -57,31 +63,31 @@ func (k Keeper) SubmitTx(goCtx context.Context, msg *ictxtypes.MsgSubmitTx) (*ic
...
 		return nil, sdkerrors.Wrap(channeltypes.ErrChannelCapabilityNotFound, "failed to GetCapability")
 	}
 
 	sdkMsgs, err := msg.GetTxMsgs()
 	if err != nil {
-		k.Logger(ctx).Error("failed to GetTxMsgs", "connection_id", msg.ConnectionId, "port_id", portID, "channel_id", channelID)
+		k.Logger(ctx).Error("SubmitTx failed: failed to GetTxMsgs", "connection_id", msg.ConnectionId, "port_id", portID, "channel_id", channelID)
```
Comment (zavgorodnii) [inline x/interchaintxs/keeper/msg_server.go]: 100% based on user input, DEBUG

**Review by zavgorodnii — CHANGES_REQUESTED** [review]
please use ERROR level only for errors that are not triggered by user input

File: x/interchainqueries/keeper/abci.go [hunk 11]
```diff
@@ -1,21 +1,24 @@
...
 func (k Keeper) EndBlocker(ctx sdk.Context) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
+
 	events := sdk.Events{}
 
 	// emit events for periodic queries
 	k.IterateRegisteredQueries(ctx, func(_ int64, registeredQuery types.RegisteredQuery) (stop bool) {
 		if uint64(ctx.BlockHeight()) >= registeredQuery.LastEmittedHeight+registeredQuery.UpdatePeriod {
-			k.Logger(ctx).Info("Interchainquery event emitted", "id", registeredQuery.Id)
+			k.Logger(ctx).Info("EndBlocker: interchainquery event emitted", "id", registeredQuery.Id)
```
Comment (zavgorodnii) [inline x/interchainqueries/keeper/abci.go]: I believe that this should be a Debug log

**Review by zavgorodnii — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Pull request merged**